README
Z80~Ex

ZiLOG Z80 CPU emulator
~~~~~~~~~~~~~~~~~~~~~~

Features:
---------

- precise opcode emulation (documented & undocumented)
- exact timings for each opcode (including I/O operations)
- full support for all interrupt modes
- ability to create several virtual CPUs
- portable: written in pure ANSI C
- builds as a library with simple callback-based API
- disassembler included  

Building:
--------------------------------------------------------

If you only want to use this library as a part of an emulator or other 
project and link it statically and your project uses CMake (recommended)
then skip this section and go directly to "Usage".

CMake (http://www.cmake.org/) is required to build the library.
If you use a popular Linux distribuion chances are that you already have 
it if not installed than in the stock repository.
Out-of-source-tree building is recommended.

To build on Unix and Unix-like environments (GCC,mingw,cygwin,DJGPP):
    mkdir build
    cd build
    cmake ..
To build everything (static and shared z80ex and z80ex_dasm libraries):
    gmake
To build only specific libraries and flavors:
    gmake z80ex (shared z80ex)
    gmake z80ex-static (static z80ex)
    gmake z80ex_dasm (shared dasm)
    gmake z80ex_dasm-static (static z80ex_dasm)

To build using MSVC (Visual Studio):
    - create a folder inside the source tree (e.g. "build")
    - start cmd.exe and go to the created folder and issue:
        cmake ..
    - double-click z80ex.sln to open it in Visual Studio
    - build ALL_BUILD or only necessary target(s) in Visual Studio
    (choose Release/Debug configuration as required)

You may override some build parameters by using ccmake or cmake-gui or other means CMake provides (see CMake documentation and relevant man pages):
* OPSTEP_FAST_AND_ROUGH -- fast and rough opcode step emulation mode (0 - off, 1 - on)
	when this mode is on, timings of internal I/O operations are ignored,
	and tstate callback feature is disabled


Installing (UNIX/MacOS X):
--------------------------

issue "gmake install" as superuser.
(default install prefix is /usr/local, which may be changed using ccmake or 
cmake-gui)
then do "/sbin/ldconfig" as superuser to update DSO links and cache


Usage:
------

emulator:

include <z80ex/z80ex.h> in your sources,
link with "libz80ex" (-lz80ex).

for rather subtle API documentation see "z80ex.h".

disassembler:

include <z80ex/z80ex_dasm.h> in your sources,
link with "libz80ex_dasm" (-lz80ex_dasm).

for API documentation see "z80ex_dasm.h".
also look at the "dasm.c" from the "examples" directory

If your project uses CMake here is an example (copy these lines into your
project's CMakeLists.txt):

    set (Z80EX_PATH "${PROJECT_SOURCE_DIR}/../z80ex"
             CACHE PATH "Path to Z80Ex library")
    include_directories("${Z80EX_PATH}/include")
    set (Z80EX_BINARY_DIR "${PROJECT_BINARY_DIR}/lib_z80ex")
    make_directory (${Z80EX_BINARY_DIR})
    set (Z80EX_STATIC_ONLY true)
    add_subdirectory (${Z80EX_PATH} ${Z80EX_BINARY_DIR})
    link_directories (${Z80EX_BINARY_DIR})
    target_link_libraries (my_cool_emulator_exexutable z80ex-static z80ex_dasm-static)

_____________________
yours, Boo-boo
boo_boo(^at^)inbox.ru


Changelog
24.08.2010 - 1.1.19pre1
- cmake build support [mkoloberdin]
- fixes for MSVC compilation [mkoloberdin]

12.09.2009 - 1.1.18
- HALT/NMI bugfix (by Kalvis)
- some code were optimized for speed (by Mastermind)
- RETI callback were implemented (by Torsten Paul)

04.06.2009 - 1.1.17
- ability to fetch z80ex version info at runtime, see z80ex_get_version()
- makefile fixes and cleanups, OS X support
- versioning scheme changed

28.09.2008 - 0.16.2
- made build for Symbian possible (and maybe for other environments
  which does not support the use of writable static data in DLL's) (thanks to ander)
- build option for fast and rough opcode timings emulation, see README

31.07.2008 - 0.16.1
- added -fPIC option to compile flags (thanks to ander)

16.07.2008 - 0.16
- LD A,R / LD A,I quirk implemented (If a LD A,I or LD A,R (which copy IFF2 to the P/V flag) is interrupted,
  then the P/V flag is reset, even if interrupts were enabled beforehand)(thanks to Restorer)
- NMI IFF2=IFF1 fix

03.04.2006 - 0.15 (major bugfix release)
- nasty deadlock in IM0 mode was eliminated

29.03.2006 - 0.14
- full MEMPTR (BIT n,(HL) flags) support!
  it seems that all undocumented z80 features are emulated now.
- undocumented flags were fixed for bit n,reg8

17.03.2006 - 0.14pre1
- massive code reorganisation
- include files are now installed under include/z80ex
- disassembler added

15.03.2006 - release 0.13
- more MEMPTR emulation
- fixed nasty bug in DD/FD prefixes behavior

02.03.2006
- fixed IN A,(nn) (flags)

23.02.2006 - release 0.12
- fixed timings for LD IIh/IIl,nn (thanks to Restorer)

22.02.2006
- fixed timings for IIh/IIl undocumented opcodes

18.02.2006
- fixed undocumented flag effects for BIT n,(IX+d)
- undocumented flag effects implemented for block I/O operations 
- undocumented internal register MEMPTR emulation (incomplete, more info needed) 

16.02.2006
- fixed sized data types declaration for MSVC (thanks to SMT)

15.02.2006 - release 0.11
- added M1 signal state to memory read callback parameters
- DAA emulation has been fixed (now uses table from SMT's UnrealSpeccy)

11.02.2006
- minor API changes

10.02.2006 - release 0.1


z80ex.c
/*
 * Z80~Ex, ZILoG Z80 CPU emulator.
 *
 * by Boo-boo [boo_boo(^at^)inbox.ru]
 * contains code from the FUSE project (http://fuse-emulator.sourceforge.net)
 * released under GNU GPL v2
 *
 */

#include <stdlib.h>
#include <string.h>

#define __Z80EX_SELF_INCLUDE

#include "typedefs.h"
#include "z80ex.h"
#include "macros.h"

#define temp_byte cpu->tmpbyte
#define temp_byte_s cpu->tmpbyte_s
#define temp_addr cpu->tmpaddr
#define temp_word cpu->tmpword

#include "ptables.c"
#include "opcodes/opcodes_base.c"
#include "opcodes/opcodes_dd.c"
#include "opcodes/opcodes_fd.c"
#include "opcodes/opcodes_cb.c"
#include "opcodes/opcodes_ed.c"
#include "opcodes/opcodes_ddcb.c"
#include "opcodes/opcodes_fdcb.c"

/*#define _DOQUOTE(x) #x
#define DOQUOTE(x) _DOQUOTE(x)*/

#define DOQUOTE(x) #x

static char revision_type[]=DOQUOTE(Z80EX_RELEASE_TYPE);
static char ver_str[]=DOQUOTE(Z80EX_VERSION_STR);
static Z80EX_VERSION version = {Z80EX_API_REVISION, Z80EX_VERSION_MAJOR, Z80EX_VERSION_MINOR, revision_type, ver_str};

LIB_EXPORT Z80EX_VERSION *z80ex_get_version()
{
	return(&version);
}

/* do one opcode (instruction or prefix) */
LIB_EXPORT int z80ex_step(Z80EX_CONTEXT *cpu)
{
	Z80EX_BYTE opcode, d;
	z80ex_opcode_fn ofn=NULL;
	
	cpu->doing_opcode=1;
	cpu->noint_once=0;
	cpu->reset_PV_on_int=0;
	cpu->tstate=0;
	cpu->op_tstate=0;
	
	opcode=READ_OP_M1(); /*fetch opcode*/
	if(cpu->int_vector_req)
	{
		TSTATES(2); /*interrupt eats two extra wait-states*/
	}
	R++; /*R increased by one on every first M1 cycle*/

	T_WAIT_UNTIL(4); /*M1 cycle eats min 4 t-states*/

	if(!cpu->prefix) opcodes_base[opcode](cpu);
	else
	{
		if((cpu->prefix | 0x20) == 0xFD && ((opcode | 0x20) == 0xFD || opcode == 0xED))
		{
			cpu->prefix=opcode;
			cpu->noint_once=1; /*interrupts are not accepted immediately after prefix*/
		}
		else
		{
			switch(cpu->prefix)
			{
				case 0xDD:
				case 0xFD:
					if(opcode == 0xCB)
					{
						d=READ_OP(); /*displacement*/
						temp_byte_s=(d & 0x80)? -(((~d) & 0x7f)+1): d;
						opcode=READ_OP();
						ofn = (cpu->prefix == 0xDD)? opcodes_ddcb[opcode]: opcodes_fdcb[opcode];						
					}
					else
					{
						ofn = (cpu->prefix == 0xDD)? opcodes_dd[opcode]: opcodes_fd[opcode];
						if(ofn == NULL) ofn=opcodes_base[opcode]; /*'mirrored' instructions*/
					}
					break;
								
				case 0xED:
					ofn = opcodes_ed[opcode];
					if(ofn == NULL) ofn=opcodes_base[0x00];
					break;
				
				case 0xCB:
					ofn = opcodes_cb[opcode];
					break;
					
				default:
					/*this must'nt happen!*/
					break;
			}
		
			ofn(cpu);
		
			cpu->prefix=0;
		}
	}
		
	cpu->doing_opcode=0;
	return(cpu->tstate);
}

LIB_EXPORT Z80EX_BYTE z80ex_last_op_type(Z80EX_CONTEXT *cpu)
{
	return(cpu->prefix);
}
	
LIB_EXPORT void z80ex_reset(Z80EX_CONTEXT *cpu)
{
	PC=0x0000; IFF1=IFF2=0; IM=IM0;
	AF=SP=BC=DE=HL=IX=IY=AF_=BC_=DE_=HL_=0xffff;
	I=R=R7=0;
	cpu->noint_once=0; cpu->reset_PV_on_int=0; cpu->halted=0;
	cpu->int_vector_req=0;
	cpu->doing_opcode=0;
	cpu->tstate=cpu->op_tstate=0;
	cpu->prefix=0;
}

/**/
LIB_EXPORT Z80EX_CONTEXT *z80ex_create(
	z80ex_mread_cb mrcb_fn, void *mrcb_data,
	z80ex_mwrite_cb mwcb_fn, void *mwcb_data,
	z80ex_pread_cb prcb_fn, void *prcb_data,
	z80ex_pwrite_cb pwcb_fn, void *pwcb_data,
	z80ex_intread_cb ircb_fn, void *ircb_data
)
{
	Z80EX_CONTEXT *cpu;
	
	if((cpu=(Z80EX_CONTEXT *)malloc(sizeof(Z80EX_CONTEXT))) == NULL) return(NULL);
	memset(cpu,0x00,sizeof(Z80EX_CONTEXT));
	
	z80ex_reset(cpu);
	
	cpu->mread_cb=mrcb_fn;
	cpu->mread_cb_user_data=mrcb_data;
	cpu->mwrite_cb=mwcb_fn;
	cpu->mwrite_cb_user_data=mwcb_data;	
	cpu->pread_cb=prcb_fn;
	cpu->pread_cb_user_data=prcb_data;	
	cpu->pwrite_cb=pwcb_fn;
	cpu->pwrite_cb_user_data=pwcb_data;
	cpu->intread_cb=ircb_fn;
	cpu->intread_cb_user_data=ircb_data;	
	
	return(cpu);
}

LIB_EXPORT void z80ex_destroy(Z80EX_CONTEXT *cpu)
{
	free(cpu);
}

LIB_EXPORT void z80ex_set_tstate_callback(Z80EX_CONTEXT *cpu, z80ex_tstate_cb cb_fn, void *user_data)
{
	cpu->tstate_cb=cb_fn;
	cpu->tstate_cb_user_data=user_data;
}

LIB_EXPORT void z80ex_set_reti_callback(Z80EX_CONTEXT *cpu, z80ex_reti_cb cb_fn, void *user_data)
{
	cpu->reti_cb=cb_fn;
	cpu->reti_cb_user_data=user_data;
}

/*non-maskable interrupt*/
LIB_EXPORT int z80ex_nmi(Z80EX_CONTEXT *cpu)
{
	if(cpu->doing_opcode || cpu->noint_once || cpu->prefix) return(0);	
	
	if(cpu->halted) { PC++; cpu->halted = 0; } /*so we met an interrupt... stop waiting*/
	
	cpu->doing_opcode=1;
	
	R++; /*accepting interrupt increases R by one*/
	IFF2=IFF1;
	IFF1=0;

	TSTATES(5); 
	
	cpu->mwrite_cb(cpu, --SP, cpu->pc.b.h, cpu->mwrite_cb_user_data); /*PUSH PC -- high byte */
	TSTATES(3);
		
	cpu->mwrite_cb(cpu, --SP, cpu->pc.b.l, cpu->mwrite_cb_user_data); /*PUSH PC -- low byte */
	TSTATES(3);
	
	PC=0x0066;
	MEMPTR=PC; /*FIXME: is that really so?*/
		
	cpu->doing_opcode=0;
	
	return(11); /*NMI always takes 11 t-states*/
}

/*maskable interrupt*/
LIB_EXPORT int z80ex_int(Z80EX_CONTEXT *cpu)
{
	Z80EX_WORD inttemp;
	Z80EX_BYTE iv;
	unsigned long tt;
	
	/*If the INT line is low and IFF1 is set, and there's no opcode executing just now,
	a maskable interrupt is accepted, whether or not the
	last INT routine has finished*/
	if(!IFF1 || cpu->noint_once || cpu->doing_opcode || cpu->prefix) return(0);

	cpu->tstate=0;
	cpu->op_tstate=0;
	
	if(cpu->halted) { PC++; cpu->halted = 0; } /*so we met an interrupt... stop waiting*/
	
	/*When an INT is accepted, both IFF1 and IFF2 are cleared, preventing another interrupt from
	occurring which would end up as an infinite loop*/
	IFF1=IFF2=0;

	/*original (NMOS) zilog z80 bug:*/
	/*If a LD A,I or LD A,R (which copy IFF2 to the P/V flag) is interrupted, then the P/V flag is reset, even if interrupts were enabled beforehand.*/
	/*(this bug was fixed in CMOS version of z80)*/
	if(cpu->reset_PV_on_int) {F = (F & ~FLAG_P);}
	cpu->reset_PV_on_int=0;

	cpu->int_vector_req=1;
	cpu->doing_opcode=1;
	
	switch(IM)
	{
		case IM0:
			/*note: there's no need to do R++ and WAITs here, it'll be handled by z80ex_step*/
			tt=z80ex_step(cpu);
		
			while(cpu->prefix) /*this is not the end?*/
			{
				tt+=z80ex_step(cpu);
			}
			
			cpu->tstate=tt;
			break;
		
		case IM1:
			R++; 
			TSTATES(2); /*two extra wait-states*/
			/*An RST 38h is executed, no matter what value is put on the bus or what
			value the I register has. 13 t-states (2 extra + 11 for RST).*/
			opcodes_base[0xff](cpu); /*RST38*/
			break;
		
		case IM2:
			R++; 
			/*takes 19 clock periods to complete (seven to fetch the
			lower eight bits from the interrupting device, six to save the program
			counter, and six to obtain the jump address)*/
			iv=READ_OP();
			T_WAIT_UNTIL(7);
			inttemp=(0x100*I)+iv;
		
			PUSH(PC,7,10);
				
			READ_MEM(PCL,inttemp++,13); READ_MEM(PCH,inttemp,16);
			MEMPTR=PC;
			T_WAIT_UNTIL(19);
			
			break;
	}
	
	cpu->doing_opcode=0;
	cpu->int_vector_req=0;
	
	return(cpu->tstate);
}

LIB_EXPORT void z80ex_w_states(Z80EX_CONTEXT *cpu, unsigned w_states)
{
	TSTATES(w_states);
}

LIB_EXPORT void z80ex_next_t_state(Z80EX_CONTEXT *cpu)
{
	if(cpu->tstate_cb != NULL) cpu->tstate_cb(cpu, cpu->tstate_cb_user_data);
	cpu->tstate++;
	cpu->op_tstate++;
}

LIB_EXPORT Z80EX_WORD z80ex_get_reg(Z80EX_CONTEXT *cpu, Z80_REG_T reg)
{
	switch(reg)
	{
		case regAF: return(AF);
		case regBC: return(BC);
		case regDE: return(DE);
		case regHL: return(HL);
		case regAF_: return(AF_);
		case regBC_: return(BC_);
		case regDE_: return(DE_);
		case regHL_: return(HL_);
		case regIX: return(IX);
		case regIY: return(IY);
		case regPC: return(PC);
		case regSP: return(SP);
		case regI: return(I);
		case regR: return(R);
		case regR7: return(R7);	
		case regIM: return(IM);
		case regIFF1: return(IFF1);
		case regIFF2: return(IFF2);
	}
	
	return(0);
}

LIB_EXPORT void z80ex_set_reg(Z80EX_CONTEXT *cpu, Z80_REG_T reg, Z80EX_WORD value)
{
	switch(reg)
	{
		case regAF: AF=value; return;
		case regBC: BC=value; return;
		case regDE: DE=value; return;
		case regHL: HL=value; return;
		case regAF_: AF_=value; return;
		case regBC_: BC_=value; return;
		case regDE_: DE_=value; return;
		case regHL_: HL_=value; return;
		case regIX: IX=value; return;
		case regIY: IY=value; return;
		case regPC: PC=value; return;
		case regSP: SP=value; return;
		case regI: I=(value & 0xff); return;
		case regR: R=(value & 0xff); return;
		case regR7: R7=(value & 0xff); return;	
		case regIM:
			switch(value & 0x03)
			{
				case 0: IM=IM0; return;
				case 1: IM=IM1; return;
				case 2: IM=IM2; return;
			}
		case regIFF1: IFF1=(value & 0x01); return;
		case regIFF2: IFF2=(value & 0x01); return;
	}
	
	return;
}

LIB_EXPORT int z80ex_op_tstate(Z80EX_CONTEXT *cpu)
{
	return(cpu->tstate);
}

LIB_EXPORT int z80ex_doing_halt(Z80EX_CONTEXT *cpu)
{
	return(cpu->halted);
}

/*LIB_EXPORT int z80ex_get_noint_once(Z80EX_CONTEXT *cpu)
{
	return(cpu->noint_once);
}*/

LIB_EXPORT int z80ex_int_possible(Z80EX_CONTEXT *cpu)
{
	return((!IFF1 || cpu->noint_once || cpu->doing_opcode || cpu->prefix)? 0 : 1);
}

LIB_EXPORT int z80ex_nmi_possible(Z80EX_CONTEXT *cpu)
{
	return((cpu->noint_once || cpu->doing_opcode || cpu->prefix)? 0 : 1);
}



include/z80ex.h
/*
 * Z80~Ex, ZILoG Z80 CPU emulator.
 *
 * by Boo-boo [boo_boo(^at^)inbox.ru]
 * contains code from the FUSE project (http://fuse-emulator.sourceforge.net)
 * Released under GNU GPL v2
 *
 */

#ifndef _Z80EX_H_INCLUDED
#define _Z80EX_H_INCLUDED

#include "z80ex_common.h"

typedef
enum {regAF,regBC,regDE,regHL,regAF_,regBC_,regDE_,regHL_,regIX,regIY,regPC,regSP,regI,regR,regR7,regIM/*0,1 or 2*/,regIFF1,regIFF2}
Z80_REG_T;

typedef struct {
	int API_revision;
	int major;
	int minor;
	char *release_type; /*e.g., "beta", "RC"*/
	char *as_string; /*full version string, e.g., "0.16.7beta"*/
} Z80EX_VERSION;

#ifndef __Z80EX_SELF_INCLUDE

struct _z80_cpu_context;
typedef struct _z80_cpu_context Z80EX_CONTEXT;

#endif

/*callback prototypes:*/

/*to be called on each T-State [optional]*/
typedef void (*z80ex_tstate_cb)(Z80EX_CONTEXT *cpu, void *user_data);

/*read byte from memory <addr> -- called when RD & MREQ goes active.
m1_state will be 1 if M1 signal is active*/
typedef Z80EX_BYTE (*z80ex_mread_cb)(Z80EX_CONTEXT *cpu, Z80EX_WORD addr, int m1_state, void *user_data);

/*write <value> to memory <addr> -- called when WR & MREQ goes active*/
typedef void (*z80ex_mwrite_cb)(Z80EX_CONTEXT *cpu, Z80EX_WORD addr, Z80EX_BYTE value, void *user_data);

/*read byte from <port> -- called when RD & IORQ goes active*/
typedef Z80EX_BYTE (*z80ex_pread_cb)(Z80EX_CONTEXT *cpu, Z80EX_WORD port, void *user_data);

/*write <value> to <port> -- called when WR & IORQ goes active*/
typedef void (*z80ex_pwrite_cb)(Z80EX_CONTEXT *cpu, Z80EX_WORD port, Z80EX_BYTE value, void *user_data);

/*read byte of interrupt vector -- called when M1 and IORQ goes active*/
typedef Z80EX_BYTE (*z80ex_intread_cb)(Z80EX_CONTEXT *cpu, void *user_data);


#ifndef __Z80EX_SELF_INCLUDE
	
#ifdef __cplusplus
extern "C" {
#endif

/*get version info*/
extern Z80EX_VERSION *z80ex_get_version();

/*create and initialize CPU*/
extern Z80EX_CONTEXT *z80ex_create(z80ex_mread_cb mrcb_fn, void *mrcb_data,
	z80ex_mwrite_cb mwcb_fn, void *mwcb_data,
	z80ex_pread_cb prcb_fn, void *prcb_data,
	z80ex_pwrite_cb pwcb_fn, void *pwcb_data,
	z80ex_intread_cb ircb_fn, void *ircb_data);

/*destroy CPU*/
extern void z80ex_destroy(Z80EX_CONTEXT *cpu);

/*do next opcode (instruction or prefix), return number of T-states*/
extern int z80ex_step(Z80EX_CONTEXT *cpu);

/*returns type of the last opcode, processed with z80ex_step.
type will be 0 for complete instruction, or dd/fd/cb/ed for opcode prefix.*/
extern Z80EX_BYTE z80ex_last_op_type(Z80EX_CONTEXT *cpu);
	
/*set T-state callback*/
extern void z80ex_set_tstate_callback(Z80EX_CONTEXT *cpu, z80ex_tstate_cb cb_fn, void *user_data);

/*maskable interrupt*/
/*returns number of T-states if interrupt was accepted, otherwise 0*/
extern int z80ex_int(Z80EX_CONTEXT *cpu);

/*non-maskable interrupt*/
/*returns number of T-states (11 if interrupt was accepted, or 0 if processor
is doing an instruction right now)*/
extern int z80ex_nmi(Z80EX_CONTEXT *cpu);

/*reset CPU*/
extern void z80ex_reset(Z80EX_CONTEXT *cpu);

/*get register value*/
extern Z80EX_WORD z80ex_get_reg(Z80EX_CONTEXT *cpu, Z80_REG_T reg);

/*set register value (for 1-byte registers lower byte of <value> will be used)*/
extern void z80ex_set_reg(Z80EX_CONTEXT *cpu, Z80_REG_T reg, Z80EX_WORD value);

/*returns 1 if CPU doing HALT instruction now*/
extern int z80ex_doing_halt(Z80EX_CONTEXT *cpu);

/*when called from callbacks, returns current T-state of the executing opcode (instruction or prefix),
else returns T-states taken by last opcode executed*/
extern int z80ex_op_tstate(Z80EX_CONTEXT *cpu);

/*generate <w_states> Wait-states. (T-state callback will be called <w_states> times, when defined).
should be used to simulate WAIT signal or disabled CLK*/ 
extern void z80ex_w_states(Z80EX_CONTEXT *cpu, unsigned w_states);

/*spend one T-state doing nothing (often IO devices cannot handle data request on
the first T-state at which RD/WR goes active).
for I/O callbacks*/
extern void z80ex_next_t_state(Z80EX_CONTEXT *cpu);

/*returns 1 if maskable interrupts are possible in current z80 state*/
extern int z80ex_int_possible(Z80EX_CONTEXT *cpu);

/*returns 1 if non-maskable interrupts are possible in current z80 state*/
extern int z80ex_nmi_possible(Z80EX_CONTEXT *cpu);

#ifdef __cplusplus
}
#endif

#endif

#endif


z80ex_dasm.c
/*
 * Z80~Ex, ZILoG Z80 CPU emulator.
 *
 * by Boo-boo [boo_boo(^at^)inbox.ru]
 * contains code from the FUSE project (http://fuse-emulator.sourceforge.net)
 * Released under GNU GPL v2
 *
 */
 
#include <stdlib.h>
#include <string.h>
#define __USE_ISOC99
#include <stdio.h>
#include <stdarg.h>
 
#define __Z80EX_SELF_INCLUDE
#include "z80ex_dasm.h"

typedef struct {
	const char *mnemonic;
	int t_states;
	int t_states2;
} z80ex_opc_dasm;

#include "opcodes/opcodes_dasm.c"

static const char *formats[] = {
	"#%02X", /*bytes*/	
	"#%04X", /*words*/
	"%d" /*WORDS_DEC, BYTES_DEC*/
};

#define STMP_SIZE 100

LIB_EXPORT int z80ex_dasm(char *output, int output_size, unsigned flags, int *t_states, int *t_states2,
	z80ex_dasm_readbyte_cb readbyte_cb, Z80EX_WORD addr, void *user_data)
{
	Z80EX_BYTE opc=0, next=0, disp_u=0;
	Z80EX_SIGNED_BYTE disp;
	int have_disp=0;
	int out_len=0;
	int bytes=0;
	const char *bytes_format=formats[0];
	const char *words_format=formats[1];
	const z80ex_opc_dasm *dasm = NULL;
	static char stmp[STMP_SIZE];
	
	if(flags & WORDS_DEC) words_format = formats[2];
	if(flags & BYTES_DEC) bytes_format = formats[2];
	
	*output='\0';
	*t_states=0;
	*t_states2=0;
	
	opc = readbyte_cb(addr++,user_data);
	bytes++;

	switch(opc)
	{
		case 0xDD:
		case 0xFD:
			next = readbyte_cb(addr++,user_data);
			if((next | 0x20) == 0xFD || next == 0xED)
			{
				strncpy(output,"NOP*",output_size-1);
				*t_states=4;
				dasm=NULL;
			}
			else if(next == 0xCB)
			{
				disp_u = readbyte_cb(addr++,user_data);
				next = readbyte_cb(addr++,user_data);
				bytes+=3;
				
				dasm = (opc==0xDD)? &dasm_ddcb[next]: &dasm_fdcb[next];
				have_disp=1;
			}
			else
			{
				bytes++;
				dasm = (opc==0xDD)? &dasm_dd[next]: &dasm_fd[next];
				if(dasm->mnemonic == NULL) /*mirrored instructions*/
				{
					dasm = &dasm_base[next];
					*t_states=4;
					*t_states2=4;
				}
			}
			break;
			
		case 0xED:
			next = readbyte_cb(addr++,user_data);
			bytes++;
			dasm = &dasm_ed[next];
			if(dasm->mnemonic == NULL)
			{
				strncpy(output,"NOP*",output_size-1);
				*t_states=8;
				dasm=NULL;
			}
			break;
			
		case 0xCB:
			next = readbyte_cb(addr++,user_data);
			bytes++;
			dasm = &dasm_cb[next];
			break;
		
		default:
			dasm = &dasm_base[opc];
			break;
	}
	
	if(dasm!=NULL)
	{	
		const char *mpos;
		int arglen;
		Z80EX_BYTE hi,lo; 
		char *outpos=output;
		
		for(mpos=(dasm->mnemonic); *mpos && out_len < output_size; mpos++)
		{
			*stmp='\0';
			
			switch(*mpos)
			{
				case '@':
					lo=readbyte_cb(addr++,user_data);
					hi=readbyte_cb(addr++,user_data);
					bytes+=2;
				
					arglen=snprintf(stmp,STMP_SIZE,words_format,(int)(lo+hi*0x100));
				
					break;
				
				case '$':
				case '%':
					if(!have_disp) disp_u = readbyte_cb(addr++,user_data);
					bytes++;
					disp = (disp_u & 0x80)? -(((~disp_u) & 0x7f)+1): disp_u;				
					
					if(*mpos == '$')
						arglen=snprintf(stmp,STMP_SIZE,bytes_format,(int)disp);
					else
						arglen=snprintf(stmp,STMP_SIZE,words_format,(int)((Z80EX_WORD)(addr+disp)));

					break;
				
				case '#':
					lo = readbyte_cb(addr++,user_data);
					bytes++;
				
					arglen=snprintf(stmp,STMP_SIZE,bytes_format,(int)lo);
				
					break;
				
				default:
					*(outpos++) = *mpos;
					out_len++;
					arglen=0;
					break;
			}
			
			if(arglen)
			{
				if(out_len+arglen >= output_size) break;
				strcpy(outpos,stmp);
				out_len+=arglen;
				outpos+=arglen;
			}
		}
		
		*outpos = '\0';
		
		*t_states+=dasm->t_states;
		*t_states2+=dasm->t_states2;
	}
	
	if(*t_states == *t_states2) *t_states2=0;
	
	return(bytes);
}


